src/measure_theory/decomposition/signed_hahn.lean

```diff
@@ -0,0 +1,590 @@
+Copyright (c) 2021 Kexing Ying. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kexing Ying
+-/
+import measure_theory.vector_measure
+import order.symm_diff
+
+/-!
+# Hahn decomposition
+
+This file defines positive and negative sets with respect to a vector measure and
+prove the Hahn decomposition (signed version).
+
+## Main definitions
+
+* `vector_measure.positive` : a set `i` is positive with respect to a vector measure if for all
+  measurable sets `j` contained in `i`, the measure of `j` is non-negative.
+* `vector_measure.negative` : a set `i` is negative with respect to a vector measure if for all
+  measurable sets `j` contained in `i`, the measure of `j` is non-positive.
+
+## Main results
+
+* `signed_measure.exists_disjoint_positive_negative_union_eq` : the Hahn decomposition theorem.
+* `signed_measure.exists_negative_set` : A measurable set of negative measure contains at least
+  one negative subset.
+
+## Tags
+
+positive, negative, Hahn decomposition
+-/
+
+noncomputable theory
+open_locale classical big_operators nnreal ennreal
+
+section lemmas
+
+variables {α : Type*}
+
+open set filter
+
+lemma exists_tendsto_Inf {S : set ℝ} (hS : ∃ x, x ∈ S) (hS' : ∃ x, ∀ y ∈ S, x ≤ y) :
+  ∃ (f : ℕ → ℝ) (hf : ∀ n, f n ∈ S), tendsto f at_top (nhds (Inf S)) :=
+begin
+  have : ∀ n : ℕ, ∃ t ∈ S, t < Inf S + 1 / (n + 1 : ℝ),
+  { exact λ n, (real.Inf_lt _ hS hS').1 ((lt_add_iff_pos_right _).2 nat.one_div_pos_of_nat) },
+  choose f hf using this,
+  refine ⟨f, λ n, (hf n).1, _⟩,
+  rw tendsto_iff_dist_tendsto_zero,
+  refine squeeze_zero_norm _ tendsto_one_div_add_at_top_nhds_0_nat,
+  intro n,
+  obtain ⟨hf₁, hf₂⟩ := hf n,
+  rw [real.dist_eq, real.norm_eq_abs, abs_abs,
+      abs_of_nonneg (sub_nonneg.2 (real.Inf_le S hS' hf₁))],
+  linarith,
+end
+
+lemma exists_tendsto_Sup {S : set ℝ} (hS : ∃ x, x ∈ S) (hS' : ∃ x, ∀ y ∈ S, y ≤ x) :
+  ∃ (f : ℕ → ℝ) (hf : ∀ n, f n ∈ S), tendsto f at_top (nhds (Sup S)) :=
+begin
+  have : ∀ n : ℕ, ∃ t ∈ S, Sup S - 1 / (n + 1 : ℝ) < t,
+  { intro n,
+    apply (real.lt_Sup _ hS hS').1,
+    rw [sub_lt, _root_.sub_self],
+    exact nat.one_div_pos_of_nat },
+  choose f hf using this,
+  refine ⟨f, λ n, (hf n).1, _⟩,
+  rw tendsto_iff_dist_tendsto_zero,
+  refine squeeze_zero_norm _ tendsto_one_div_add_at_top_nhds_0_nat,
+  intro n,
+  obtain ⟨hf₁, hf₂⟩ := hf n,
+  rw [real.dist_eq, real.norm_eq_abs, abs_abs, abs_of_nonpos],
+  { linarith },
+  { rw [sub_le, _root_.sub_zero],
+    exact real.le_Sup S hS' hf₁ },
+end
+
+end lemmas
+
+variables {α β : Type*} [measurable_space α]
+variables {M : Type*} [add_comm_monoid M] [topological_space M] [ordered_add_comm_monoid M]
+
+namespace measure_theory
+
+namespace vector_measure
+
+/-- A set `i` is positive with respect to a vector measure if for all
+measurable set `j`, `j ⊆ i`, `j` has non-negative measure. -/
+def positive (v : vector_measure α M) (i : set α) : Prop :=
+∀ ⦃j⦄, j ⊆ i → measurable_set j → 0 ≤ v j
```

Comment:
The `measurable_set j →` condition is unecessary right, since `v 0` is defined as `0`?

Reply:
Right, it's not necessary but it is nice to have when proving. I think if needed I can take the `absolutely_continuous` approach where we don't require measurable set in the definition but make a `mk` lemma to include the condition

Reply:
I've added the lemmas `subset_nonneg_of_positive` and `positive_of_subset_nonneg` which are the stronger version for, using and proving respectively